File: tests/types_tutorial/stage3_closed_sets/payments.ts
// The strings become closed sets, and each provider's vocabulary stops at the door.

declare const brand: unique symbol;
type Branded<T, B> = T & { readonly [brand]: B };

export type PaymentId = Branded<string, "PaymentId">;
export type CustomerId = Branded<string, "CustomerId">;
export type VendorId = Branded<string, "VendorId">;

export const RETRY_LIMIT: number = 3;

export type Provider = "stripe" | "adyen" | "wise";

export type PaymentState = "authorized" | "captured" | "refunded" | "failed";

const PROVIDER_VOCABULARY: Record<Provider, Record<string, PaymentState>> = {
  stripe: {
    requires_capture: "authorized",
    succeeded: "captured",
    canceled: "failed",
  },
  adyen: {
    Authorised: "authorized",
    SettleScheduled: "captured",
    Refused: "failed",
  },
  wise: {
    pending: "authorized",
    completed: "captured",
    failed: "failed",
  },
};

/** A word this provider does not use is not a state; it is unread input. */
export function providerState(
  provider: Provider,
  raw: string,
): PaymentState | undefined {
  return PROVIDER_VOCABULARY[provider][raw];
}

export function findPayment(
  payments: Record<string, object>,
  paymentId: PaymentId,
): object | undefined {
  return payments[paymentId];
}

export function authorize(payment: any, customer: any): object {
  if (payment.amount > 100000 && customer.kyc !== "verified") {
    return { ok: false, reason: "verification required" };
  }
  payment.state = "authorized";
  return { ok: true };
}

export function capture(payment: any, vendor: any): object {
  if (payment.state !== "authorized") {
    return { ok: false, reason: "not authorized" };
  }
  if (vendor.payouts !== "enabled") {
    return { ok: false, reason: "vendor cannot receive payouts" };
  }
  payment.state = "captured";
  payment.capturedAt = 1700000000;
  return { ok: true };
}

export function refund(payment: any, amount: number): object {
  if (payment.state !== "captured") {
    return { ok: false, reason: "not captured" };
  }
  if (amount > payment.amount) {
    return { ok: false, reason: "refund exceeds capture" };
  }
  payment.state = "refunded";
  return { ok: true };
}

File: tests/types_tutorial/stage3_closed_sets/payments.py
from enum import StrEnum
from typing import NewType

PaymentId = NewType("PaymentId", str)
CustomerId = NewType("CustomerId", str)
VendorId = NewType("VendorId", str)

RETRY_LIMIT: int = 3

class Provider(StrEnum):
    STRIPE = "stripe"
    ADYEN = "adyen"
    WISE = "wise"

class PaymentState(StrEnum):
    AUTHORIZED = "authorized"
    CAPTURED = "captured"
    REFUNDED = "refunded"
    FAILED = "failed"

PROVIDER_VOCABULARY: dict[Provider, dict[str, PaymentState]] = {
    Provider.STRIPE: {
        "requires_capture": PaymentState.AUTHORIZED,
        "succeeded": PaymentState.CAPTURED,
        "canceled": PaymentState.FAILED,
    },
    Provider.ADYEN: {
        "Authorised": PaymentState.AUTHORIZED,
        "SettleScheduled": PaymentState.CAPTURED,
        "Refused": PaymentState.FAILED,
    },
    Provider.WISE: {
        "pending": PaymentState.AUTHORIZED,
        "completed": PaymentState.CAPTURED,
        "failed": PaymentState.FAILED,
    },
}

def provider_state(provider: Provider, raw: str) -> PaymentState | None:
    return PROVIDER_VOCABULARY[provider].get(raw)

def find_payment(payments: dict, payment_id: PaymentId) -> dict | None:
    return payments.get(payment_id)

def authorize(payment: dict, customer: dict) -> dict:
    if payment["amount"] > 100000 and customer["kyc"] != "verified":
        return {"ok": False, "reason": "verification required"}
    payment["state"] = PaymentState.AUTHORIZED
    return {"ok": True}

def capture(payment: dict, vendor: dict) -> dict:
    if payment["state"] != PaymentState.AUTHORIZED:
        return {"ok": False, "reason": "not authorized"}
    if vendor["payouts"] != "enabled":
        return {"ok": False, "reason": "vendor cannot receive payouts"}
    payment["state"] = PaymentState.CAPTURED
    payment["captured_at"] = 1700000000
    return {"ok": True}

def refund(payment: dict, amount: int) -> dict:
    if payment["state"] != PaymentState.CAPTURED:
        return {"ok": False, "reason": "not captured"}
    if amount > payment["amount"]:
        return {"ok": False, "reason": "refund exceeds capture"}
    payment["state"] = PaymentState.REFUNDED
    return {"ok": True}
